more readable code (no templates) + a small addition to sol (#3535)
cpinitiative/usaco-guide: solutions/gold/cses-1665.mdx

- We start a new unfinished team with person $i$.
  - This transitions to the state $dp[i][j + 1][k - s[i]]$ (one more unfinished team, and we subtract from penalty as discussed above).

Two more things:

- $k$ in our DP array can be negative, so just add 5000 to each $k$; there can be at most $N/2 = 50$ unfinished intervals, and each of them contributes at most $-100$.
- $dp[i]$ depends only on $dp[i - 1]$, so we can drop the first dimension that
  we store (to save memory).

I believe that this is called the "open and close interval trick". See
[zscoder's blog post](https://codeforces.com/blog/entry/47764) for more
information and problems.

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

const int N = 101;
const int X = 5e3 + 1;
const int M = 1e9 + 7;
const int K = 5e3;  // K is offset to account for negatives

int n, x, s[N];     // x -> max penalty

/*
 * dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
 * for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
 * lines up with s[i] (i.e. skill of person i + 1 = s[i])
 */
ll dp1[N][X + K];  // i

> lll dp2[N][X + K];  // i + 1

Reviewer: the comment doesn't explain why there's 2 dp arrays

int main() {
	cin >> n >> x;
	for (int i = 0; i < n; i++) {
		cin >> s[i];
	}
	sort(s, s + n);
	
	dp1[0][K] = 1;  // dp[0][0][0] -> 0 people, 0 unfinished groups, 0 penalty
	for (int i = 0; i < n; i++) {
		for (int j = 0; j <= (n - i); j++) {  // at most n - i unfinished groups
			for (int k = 0; k <= x + K; k++) {
				if (!dp1[j][k]) continue;
				dp2[j][k] += dp1[j][k];  // i has their own group
				if (j) dp2[j - 1][k + s[i]] += j * dp1[j][k];  // finish group
				if (j + 1 <= n - (i + 1))
					dp2[j + 1][k - s[i]] +=
					    dp1[j][k];  // create new unfinished group
				if (j <= n - (i + 1))

> 					dp2[j][k] += j * dp1[j][k];  // extend unfinished group

Reviewer: braces here & everywhere elsle

- We start a new unfinished team with person $i$.
  - This transitions to the state $dp[i][j + 1][k - s[i]]$ (one more unfinished team, and we subtract from penalty as discussed above).

Two more things:

- $k$ in our DP array can be negative, so just add 5000 to each $k$; there can be at most $N/2 = 50$ unfinished intervals, and each of them contributes at most $-100$.
- $dp[i]$ depends only on $dp[i - 1]$, so we can drop the first dimension that
  we store (to save memory).

I believe that this is called the "open and close interval trick". See
[zscoder's blog post](https://codeforces.com/blog/entry/47764) for more
information and problems.

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

const int N = 101;
const int X = 5e3 + 1;
const int M = 1e9 + 7;
const int K = 5e3;  // K is an offset to account for negatives
int n, x, s[N];     // x -> max penalty

// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
// for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
// lines up with s[i] (i.e. skill of person i + 1 = s[i])

> ll dp1[N][X + K];  // dp[i][j][k]
> ll dp2[N][X + K];  // dp[i + 1][j][k]
> // Naively, our dp table would require N * N * X memory
> // However, dp[i][j][k] can only affect dp[i + 1][x][y],
> // so we effectively only need to store two rows of the dp table.

Reviewer: ```suggestion
// Notice we only need to store 2 rows!
ll dp1[N][X + K];  // The subarray we're currently at
ll dp2[N][X + K];  // The next subarray (we'll fill this up with dp1)
```

Reviewer: i hope my understanding of your code is correct

Author: yeah it is :)

  - This transitions to the state $dp[i][j][k]$ (again, unfinished teams and penalty are not affected). There are $j$ unfinished teams we can choose to extend, so $dp[i][j][k] \mathrel{+}= j \cdot dp[i - 1][j][k]$.
- We use person $i$ to "finish" an unfinished team.
  - This transitions to the state $dp[i][j - 1][k + s_i]$ (one less unfinished team, and we add to penalty as discussed above). There are again $j$ unfinished teams to choose from, so $dp[i][j - 1][k + s[i]] \mathrel{+}= j \cdot dp[i - 1][j][k]$.
- We start a new unfinished team with person $i$.
  - This transitions to the state $dp[i][j + 1][k - s[i]]$ (one more unfinished team, and we subtract from penalty as discussed above).

Two more things:

- $k$ in our DP array can be negative, so just add 5000 to each $k$; there can be at most $N/2 = 50$ unfinished intervals, and each of them contributes at most $-100$.
- $dp[i]$ depends only on $dp[i - 1]$, so we can drop the first dimension that
  we store (to save memory).

I believe that this is called the "open and close interval trick". See
[zscoder's blog post](https://codeforces.com/blog/entry/47764) for more
information and problems.

```cpp
#include <bits/stdc++.h>

> #define int long long

Reviewer: use a `using` declaration instead & use `ll`

using namespace std;

> const int N = 101, X = 5e3 + 1, M = 1e9 + 7,

Reviewer: declarations on separate lines

          K = 5e3;  // K is offset to account for negatives
int n, x, s[N];     // x -> max penalty
// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
// for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
// lines up with s[i] (i.e. skill of person i + 1 = s[i])

> int dp1[N][X + K], dp2[N][X + K];  // dp1 -> i, dp2 -> i + 1

Reviewer: can you just use vectors instead

Author: why vectors?

Reviewer: easier to debug

Reviewer: bump

Author: i updated it to use vectors, but i dont use them very often so im not sure if there's any standard practices im neglecting

Reviewer: nah u did good

> signed main() {

Reviewer: ```suggestion
int main() {
```

	cin >> n >> x;
	for (int i = 0; i < n; i++) cin >> s[i];
	sort(s, s + n);
	dp1[0][K] = 1;  // dp[0][0][0] -> 0 people, 0 unfinished groups, 0 penalty
	for (int i = 0; i < n; i++) {
		for (int j = 0; j <= (n - i); j++) {  // at most n - i unfinished groups
			for (int k = 0; k <= x + K; k++) {
				if (!dp1[j][k]) continue;
				dp2[j][k] += dp1[j][k];  // i has their own group
				if (j) dp2[j - 1][k + s[i]] += j * dp1[j][k];  // finish group
				if (j + 1 <= n - (i + 1))
					dp2[j + 1][k - s[i]] +=
					    dp1[j][k];  // create new unfinished group
				if (j <= n - (i + 1))
					dp2[j][k] += j * dp1[j][k];  // extend unfinished group
			}
		}
		for (int j = 0; j <= (n - (i + 1)); j++) {
			for (int k = 0; k <= x + K; k++) {
				dp1[j][k] = dp2[j][k] % M;  // i + 1 becomes the new i
				dp2[j][k] = 0;
			}
		}
	}
	int ans = 0;
	for (int i = K; i <= x + K; i++)

> 		ans += dp1[0][i],

Reviewer: braces

		    ans %=

> 		    M;  // dp[n][0][i] -> n people, 0 unfinished groups, penalty <= x

Reviewer: please don't try to squeeze multiple statements in w/ a comma

> 	cout << ans;

Reviewer: ```suggestion
	cout << ans << endl;
```

```cpp
#include <bits/stdc++.h>
using ll = long long;
using namespace std;
const int N = 101;
const int X = 5e3 + 1; 
const int M = 1e9 + 7;
const int K = 5e3;  // K is offset to account for negatives
int n, x, s[N];     // x -> max penalty
// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
// for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
// lines up with s[i] (i.e. skill of person i + 1 = s[i])
ll dp1[N][X + K]; // dp[i][j][k]
ll dp2[N][X + K]; // dp[i + 1][j][k]
// Naively, our dp table would require N * N * X memory
// However, dp[i][j][k] can only affect dp[i + 1][x][y],
// so we effectively only need to store two rows of the dp table.
int main() {
	cin >> n >> x;

> 	for (int i = 0; i < n; i++) cin >> s[i];

Reviewer: ayo why remove the braces here

Author: haha wait that's mb

```cpp
#include <bits/stdc++.h>
#define int long long
using namespace std;
const int N = 101, X = 5e3 + 1, M = 1e9 + 7,
          K = 5e3;  // K is offset to account for negatives
int n, x, t[N];     // x -> max penalty
// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty

> // contrary to the above solution, this code uses a push dp approach

Reviewer: > this code uses a push dp approach

it's very bad to do an editorial saying one method and then have code demonstrating another

Author: Thanks for the comment; I changed the sol to use push dp as well.
